kratos/processes/compute_nodal_gradient_process.cpp

```diff
@@ -74,69 +74,77 @@ void ComputeNodalGradientProcess<THistorical>::ComputeElementalContributionsAndV
         J0.resize(dimension, local_space_dimension_first_element);
 
     // Variable retriever
-    AuxiliarVariableVectorRetriever* p_variable_retriever = nullptr;
+    Kratos::unique_ptr<AuxiliarVariableVectorRetriever> p_variable_retriever;
     if (mNonHistoricalVariable) {
-        p_variable_retriever = new VariableVectorRetriever<ComputeNodalGradientProcessSettings::GetAsNonHistoricalVariable>();
+        p_variable_retriever = Kratos::make_unique<VariableVectorRetriever<ComputeNodalGradientProcessSettings::GetAsNonHistoricalVariable>>();
     } else {
-        p_variable_retriever = new VariableVectorRetriever<ComputeNodalGradientProcessSettings::GetAsHistoricalVariable>();
+        p_variable_retriever = Kratos::make_unique<VariableVectorRetriever<ComputeNodalGradientProcessSettings::GetAsHistoricalVariable>>();
     }
 
+    struct tls_type
+    {
+        Matrix DN_DX, J0, InvJ0;
+        Vector N, values;
+        double detJ0 = 0.0;
+    };
+    tls_type TLS = tls_type();
+    TLS.DN_DX = DN_DX;
+    TLS.J0 =J0;
+    TLS.InvJ0 = InvJ0;
+    TLS.N = N;
+    TLS.values = values;
+    TLS.detJ0 = detJ0;
```

Comment:
you can directly use the tls in the initialization that happens above, then you don't need to do the copy here